### PR TITLE
NixOS `networking.nat`: only port-forward traffic destined for externalIP

### PR DESCRIPTION
**Pull request by gray-heron**
## Description of changes

[networking.nat.externalIP](https://search.nixos.org/options?channel=23.05&size=50&sort=relevance&type=packages&query=externalIP) option allows to specify the exact external address to be used for NAT. This is particularly useful, if there are multiple IP addresses associated with one `networking.nat.externalInterface`.

However, before this change, even if `externalIP` was set, packets destined for *other* IP addresses on the `externalInterface` would still be captured, making it impossible to host services bound to IP addresses not meant for NAT.

## Things done

- [x] Thoroughly tested behaviour both with `externalIP` set and unset.

### CODE REVIEW

**Comment by samhug** [comment]
I'm actually not sure we want to change the existing semantics. The documentation for [`networking.nat.externalIP`](https://search.nixos.org/options?channel=unstable&show=networking.nat.externalIP) and [`networking.nat.externalIPv6`](https://search.nixos.org/options?channel=unstable&show=networking.nat.externalIPv6) imply they are intended for "outbound" SNAT, not "inbound" DNAT.

Given the existing semantics, I think you'd have to resort to `networking.nat.extraCommands` to do what you want.

If I understand correctly, this change as is would break IPv6 port forwards when `networking.nat.enableIPv6 = true` and `networking.nat.externalIP != null`.
The `mkSetupNat` function is used to generate NAT rules for both IPv4 and IPv6 but this uses `networking.nat.externalIP` for both

If we do make this change then I think we need to also update the [nftables implementation](https://github.com/NixOS/nixpkgs/blob/57a0c58e728a28bc21eed85055eac9724a08f69a/nixos/modules/services/networking/nat-nftables.nix#L82) to match

**Comment by samhug** [comment]
An alternative approach would be to add a new optional field in [`networking.nat.forwardPorts`](https://search.nixos.org/options?channel=unstable&show=networking.nat.forwardPorts) (maybe `sourceAddress` or `externalAddress`?)

I'm not sure how difficult that would be to implement

File: nixos/modules/services/networking/nat-iptables.nix [hunk 1]
```diff
@@ -60,6 +60,7 @@ let
     ${concatMapStrings (fwd: ''
       ${iptables} -w -t nat -A nixos-nat-pre \
         -i ${toString cfg.externalInterface} -p ${fwd.proto} \
+        ${optionalString (cfg.externalIP != null) "-d ${cfg.externalIP}"} \
```
Comment (mkg20001) [inline nixos/modules/services/networking/nat-iptables.nix:63]: does this require a change in nat-nftables.nix aswell?

**Comment by gray-heron** [comment]
Ah, yes, nftables needs the same change of course.

@samhug Good catch with IPv6.

> An alternative approach would be to add a new optional field in [networking.nat.forwardPorts](https://search.nixos.org/options?channel=unstable&show=networking.nat.forwardPorts) (maybe sourceAddress or externalAddress?)

Originally I've been thinking the same, but I couldn't imagine a reasonable usecase where someone needs to force a particular address for outbound connections, but at the same needs to catch inbound connections on _all_ addresses. Also with this proposal, the external address becomes very redundant:
```nix
  networking.nat = {
    enable = true;
    internalInterfaces = ["ve-+"];
    externalInterface = "enp41s0";
    externalIP = "123.123.123.123";

    forwardPorts = [
      {
        destination = "192.168.100.2:443";
        proto = "tcp";
        sourcePort = 443;
        loopbackIPs = [ "123.123.123.123" ];
        externalDestinationIP = "123.123.123.123";
      }
      {
        destination = "192.168.100.2:80";
        proto = "tcp";
        sourcePort = 80;
        loopbackIPs = [ "123.123.123.123" ];
        externalDestinationIP = "123.123.123.123";
      }
    ];
  };
 ```

> I'm not sure how difficult that would be to implement

For iptables it's trivial. For nftables tho, I will have to do some changes, i.e. remove this `fwdMap` logic https://github.com/NixOS/nixpkgs/blob/6e43623224c463088434e0b4ebd60e80e89806fb/nixos/modules/services/networking/nat-nftables.nix#L46 , and generate a separate entry for each individual port forward. Unless it's possible to have arbitrary combinations of `dport`s and `daddr`s within one entry, but I don't think so.

I'm quite new here, so I'm not sure how to wager a change in semantics vs. making it more intuitive, but I can implement it any way you decide.

Let me know if I should be fixing IPv6 in the current approach, or implement `externalDestinationIP`.

**Comment by samhug** [comment]
> I'm quite new here, so I'm not sure how to wager a change in semantics vs. making it more intuitive, but I can implement any approach you think is better.

I do think that what you're suggesting is more intuitive then how it works right now.

My thinking with that approach was that it avoids changing the behavior for existing users. After this change anyone who currently specifies `networking.nat.externalIP` will no longer have their ports forwarded from any other external IPs.
I don't know how likely it is that someone is relying on that. I just wanted to call it out so we're aware.

> So let me know if I should be fixing IPv6 in the current approach, or implement `externalDestinationIP`.

I don't actually have merge permissions here so don't let my reviews block you :smile: 
I appreciate your efforts to improve things!

**Comment by duament** [comment]
> Unless it's possible to have arbitrary combinations of dports and daddrs within one entry, but I don't think so.

It is possible 😀. You may try this

```
sudo nft add ip nixos-nat pre iifname wlp1s0 dnat ip daddr . meta l4proto . th dport map '{ 123.123.123.123 . tcp . 80 : 192.168.1.1 . 80 }'
```

**Comment by mweinelt** [comment]
I'm locking and closing this pull request, because through the rebase you inadvertently requested many maintainers for review, which subscribed them to update notifications, resulting in unnecessary spam inside everyone's inbox. While we can remove their review requests, we sadly cannot unsubscribe anyone.

Please create a new pull request and for the next time remember to set your PR to draft status before rebasing. In draft status, you can preview the list of maintainers that are about to be requested for review, which allows you to sidestep this issue.

**Pull request closed without merging**